### PR TITLE
Add file/folder permissions to Directory and Container APIs

### LINKED ISSUE #3167: configuring file permissions for `directory { withNewFile }`
### What are you trying to do?

Create a new file with static content through the API with permissions set.

### Why is this important to you?

Some tools are picky about file permissions and will fail if they're too permissive. For example, `~/.ssh/known_hosts`.

### How are you currently working around this?

_No response_

### PR DESCRIPTION
**Pull request by jlongtine**
Fixes #3167

I didn't write tests for the Container APIs because they're currently using the Directory APIs... let me know if you all think I should change that.

### CODE REVIEW

File: core/directory.go
Comment (vito) [inline core/directory.go:210]: TODone:

```suggestion
			permissions,
```

File: core/directory.go
Comment (vito) [inline core/directory.go:278]: Does this only affect the root directory?

File: core/integration/directory_test.go
Comment (vito) [inline core/integration/directory_test.go:329]: I think it'd be good to wrap this in a `t.Run("respects permissions", func(t *testing.T) {` so the test is less monolithic.

File: core/integration/directory_test.go
Comment (vito) [inline core/integration/directory_test.go:299]: Same here (`t.Run`).

**Review by vito — COMMENTED** [review]
This looks great, just some small nitpicks!

File: core/integration/directory_test.go
Comment (jlongtine) [reply core/integration/directory_test.go:329]: 👍🏻

File: core/integration/directory_test.go
Comment (jlongtine) [reply core/integration/directory_test.go:299]: 👍🏻

File: core/directory.go
Comment (jlongtine) [reply core/directory.go:278]: I'm honestly not sure. I'll add a test to check.

Comment (jlongtine) [reply core/directory.go:278]: Turns out the answer is no... it impacts all sub directories and files.

**Comment by jlongtine** [comment]
@vito I cleaned up those tests a bit, and added a few more to clarify how Mode works.

File: core/directory.go
Comment (vito) [reply core/directory.go:278]: Hm, that's unfortunate, and feels kind of surprising. Thanks for looking into it.

I wonder if we should just leave this API out instead for now. I can't think of many scenarios where I'd want to `chmod` an entire directory recursively to the same permissions. I guess `777` would make sense, but not much else.

cc @sipsma @shykes @aluzzardi for a second opinion :)

Comment (jlongtine) [reply core/directory.go:278]: Yeah, I agree about it being surprising. I'm definitely open to pulling it from `Container|Directory.withDirectory()` to reduce the element of surprise.

Comment (sipsma) [reply core/directory.go:278]: Agree recursive chmod would be surprising behavior and that we can just drop this from `withDirectory` for now to unblock.

If we did want the API to work such that it only applied to the directory and not contents, I wonder if you could do that by first doing a normal copy call without mode set and then running another copy that uses either include or exclude patterns such that the mode only applies to the directory. But that seems hacky and a lot of hoops to jump through; we can also look into upstream changes if there's no other existing options and we have a strong need for it.

**Review by sipsma — APPROVED** [review]
Think we should drop permissions from withDirectory for now but otherwise LGTM

File: core/directory.go
Comment (jlongtine) [reply core/directory.go:184]: Thanks for the catch!

**Pull request merged**
